extensions/windows-event-log/wel/WindowsEventLog.cpp

```diff
@@ -0,0 +1,200 @@
+		{
+			return;
+		}
+	}
+
+	event_timestamp_ = static_cast<PEVT_VARIANT>( rendered_values.get())[EvtSystemTimeCreated].FileTimeVal;
+
+	SYSTEMTIME st;
+	FILETIME ft;
+
+	ft.dwHighDateTime = (DWORD)((event_timestamp_ >> 32) & 0xFFFFFFFF);
+	ft.dwLowDateTime = (DWORD)(event_timestamp_ & 0xFFFFFFFF);
+
+	FileTimeToSystemTime(&ft, &st);
+	std::stringstream datestr;
+
+	std::string period = "AM";
+	auto hour = st.wHour;
+	if (hour >= 12 && hour < 24)
+		period = "PM";
+	if (hour >= 12)
+		hour -= 12;
+	datestr << st.wMonth << "/" << st.wDay << "/" << st.wYear << " " << std::setfill('0') << std::setw(2) << hour << ":" << std::setfill('0') << std::setw(2) << st.wMinute << ":" << std::setfill('0') << std::setw(2) << st.wSecond << " " << period;
+	event_timestamp_str_ = datestr.str();
+	auto level = static_cast<PEVT_VARIANT>(rendered_values.get())[EvtSystemLevel];
+	auto keyword = static_cast<PEVT_VARIANT>(rendered_values.get())[EvtSystemKeywords];
+	if (level.Type == EvtVarTypeByte) {
+		switch (level.ByteVal)
+		{
+			case WINEVENT_LEVEL_CRITICAL:
+			case WINEVENT_LEVEL_ERROR:
+				event_type_ = "Error";
+				event_type_index_ = 1;
+				break;
+			case WINEVENT_LEVEL_WARNING:
+				event_type_ = "Warning";
+				event_type_index_ = 2;
+				break;
+			case WINEVENT_LEVEL_INFO:
+			case WINEVENT_LEVEL_VERBOSE:
+				event_type_ = "Information";
+				event_type_index_ = 4;
+				break;
+			default:
+				event_type_index_ = 0;
+		};
+
+	}
+	else {
+		event_type_ = "N/A";
+	}
+
+	if (keyword.UInt64Val & WINEVENT_KEYWORD_AUDIT_SUCCESS) {
+		event_type_ = "Success Audit";
+		event_type_index_ = 8;
+	} else if (keyword.UInt64Val & EVENTLOG_AUDIT_FAILURE) {
+		event_type_ = "Failure Audit";
+		event_type_index_ = 16;
+	}
+
+	
+
+}
+
+std::string WindowsEventLogHandler::getEventMessage(EVT_HANDLE eventHandle) const
+{
+	std::string returnValue;
+	std::unique_ptr<WCHAR, utils::FreeDeleter> pBuffer;
+	DWORD dwBufferSize = 0;
+	DWORD dwBufferUsed = 0;
+	DWORD status = 0;
+
+	EvtFormatMessage(metadata_provider_, eventHandle, 0, 0, NULL, EvtFormatMessageEvent, dwBufferSize, pBuffer.get(), &dwBufferUsed);
+
+	//  we need to get the size of the buffer
+	status = GetLastError();
+	if (ERROR_INSUFFICIENT_BUFFER == status) {
+		dwBufferSize = dwBufferUsed;
+
+		/* All C++ examples use malloc and even HeapAlloc in some cases. To avoid any problems ( with EvtFormatMessage calling
+			free for example ) we will continue to use malloc and use a custom deleter with unique_ptr.
+		'*/
+		pBuffer = std::unique_ptr<WCHAR, utils::FreeDeleter>((LPWSTR)malloc(dwBufferSize * sizeof(WCHAR)));
+
+
+		if (pBuffer) {
+			EvtFormatMessage(metadata_provider_, eventHandle, 0, 0, NULL, EvtFormatMessageEvent, dwBufferSize, pBuffer.get(), &dwBufferUsed);
+		}
+		else {
+			return returnValue;
+		}
+	}
+	else if (ERROR_EVT_MESSAGE_NOT_FOUND == status || ERROR_EVT_MESSAGE_ID_NOT_FOUND == status) {
+		return returnValue;
+	}
+	else {
+		return returnValue;
```

Comment:
I see only one successful path here , where actual value is returned, all the rest are just returning empty string. 
This function should be simplified a bit but I think it would also benefit from some logging introduced.

Reply:
don't disagree @am-c-p-p can you take a look at this comment too?